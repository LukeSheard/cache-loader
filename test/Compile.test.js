import path from 'path';
import webpack from 'webpack';
import memoryfs from 'memory-fs';

/**
 * Handle tests running on different FS's
 * normalize all paths in output relative to this repos root
 */
const PROJECT_ROOT = path.join(__dirname, "..");
const replaceRoot = p => path.normalize(p.replace(PROJECT_ROOT, "~"));
const replaceDataRoots = data => ({
  ...data,
  remainingRequest: replaceRoot(data.remainingRequest)
})

describe("Test cache loader stats", () => {
  let counterKey;

  beforeEach(() => {
    counterKey = 1;
  });
  it("should write loader files correctly", (done) => {
    const compiler = webpack({
      mode: "development",
      context: __dirname,
      entry: path.join(__dirname, "__fixtures__", "entry_1.js"),
      output: {
        path: path.resolve(__dirname),
        filename: 'bundle.js',
      },
      module: {
        rules: [{
          test: /\.js$/,
          use: {
            loader: path.resolve(__dirname, '../src'),
            options: {
              cacheDirectory: "/",
              cacheKey() {
                return String(counterKey++)
              },
              read(key, callback) {
                expect(replaceRoot(key)).toMatchSnapshot();
                callback(new Error("no reading is done"));
              },
              generate(depFileName, callback) {
                const data = replaceRoot(depFileName);
                expect(data).toMatchSnapshot();
                callback(null, data);
              },
              write(key, data, callback) {

                expect({
                  key,
                  data: replaceDataRoots(data)
                }).toMatchSnapshot();
                callback();
              }
            }
          }
        }]
      }
    });

    const fs = new memoryfs();

    compiler.outputFileSystem = fs;

    compiler.run((err, stats) => {
      console.log(stats.toString());

      if (err || stats.hasErrors()) {
        throw new Error("Compile failed");
      }

      done();
    });
  });
});
