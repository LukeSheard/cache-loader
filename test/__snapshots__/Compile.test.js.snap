// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cache loader stats should write loader files correctly 1`] = `"1"`;

exports[`Test cache loader stats should write loader files correctly 2`] = `"~/test/__fixtures__/entry_1.js"`;

exports[`Test cache loader stats should write loader files correctly 3`] = `"~/src/index.js"`;

exports[`Test cache loader stats should write loader files correctly 4`] = `
Object {
  "data": Object {
    "contextDependencies": Array [],
    "dependencies": Array [
      "~/test/__fixtures__/entry_1.js",
      "~/src/index.js",
    ],
    "remainingRequest": "~/test/__fixtures__/entry_1.js",
    "result": Array [
      "require('./entry_1_import');
",
    ],
  },
  "key": "1",
}
`;

exports[`Test cache loader stats should write loader files correctly 5`] = `"2"`;

exports[`Test cache loader stats should write loader files correctly 6`] = `"~/test/__fixtures__/entry_1_import.js"`;

exports[`Test cache loader stats should write loader files correctly 7`] = `"~/src/index.js"`;

exports[`Test cache loader stats should write loader files correctly 8`] = `
Object {
  "data": Object {
    "contextDependencies": Array [],
    "dependencies": Array [
      "~/test/__fixtures__/entry_1_import.js",
      "~/src/index.js",
    ],
    "remainingRequest": "~/test/__fixtures__/entry_1_import.js",
    "result": Array [
      "console.log(\\"I'm a sub module\\");
",
    ],
  },
  "key": "2",
}
`;
